File: New Generation SmartPill/SmartPill/Model/CoreDataModels/Categories/Reminder+create.m
//
//  Reminder+create.m
//  SmartPill
//
//  Created by Dennis da Silva Nunes on 22/01/15.
//  Copyright (c) 2015 IFRN - Mobile School. All rights reserved.
//

#import "Reminder+create.h"

@implementation Reminder (create)

- (NSString*)description{
    NSString * string = [NSString stringWithFormat:@"Medicine = %@, and Reminder Schedule = %@",self.medicine.name,self.reminder_schedule.schedule];
    return string;
}

+ (Reminder *)reminderOfMedicine:(Medicine *)medicine
            withReminderSchedule:(Reminder_Schedule*)reminderSchedule
                   reminderSound:(Reminder_Sound*)reminderSound
          inManagedObjectContext:(NSManagedObjectContext *)context{
    
    Reminder * reminder = nil;
    
    if (medicine) {
        
        NSString * medicineName = medicine.name;
        
        NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Medicine"];
        request.predicate = [NSPredicate predicateWithFormat:@"name = %@", medicineName];
        
        NSError *error;
        NSArray *matches = [context executeFetchRequest:request error:&error];
        
        if (!matches) {
            
            return nil;
        } else if (![matches count]) {
            return nil;
        } else {
            for (Medicine * med in matches) {
                if ([med.name isEqualToString:medicineName]) {
                    reminder = [NSEntityDescription insertNewObjectForEntityForName:@"Reminder"
                                                             inManagedObjectContext:context];
                    reminder.medicine = med;
                    reminder.reminder_schedule = reminderSchedule;
                    reminder.reminder_sound = reminderSound;
                    
                    return reminder;
                }
            }
        }
    }
    return nil;
}

+ (void)updateOldReminder:(Reminder *)oldReminder
            toNewReminder:(Reminder*)newReminder
      fromDataBaseContext:(NSManagedObjectContext*)context{

    if (oldReminder) {
        
        NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Medicine"];
        request.predicate = [NSPredicate predicateWithFormat:@"ANY reminder = %@", oldReminder];
        
        NSError *error;
        NSArray *matches = [context executeFetchRequest:request error:&error];
        
        if (!matches) {
            NSLog(@"Error in Update Medicine");
            return;
        } else if ([matches count]) {
            for (Reminder * rem in matches) {
                if ([rem isEqual:oldReminder]) {
                    rem.reminder_schedule = newReminder.reminder_schedule;
                    if(rem.reminder_sound){rem.reminder_sound = newReminder.reminder_sound;}
                    rem.medicine = newReminder.medicine;
                }
            }
        }else if(![matches count]){
            NSLog(@"Could not find any medicine in the database");
            
        }
        
    }
    [context save:nil];
}

@end

File: SmartPill/dbModals/Categories/Reminder+Create.swift
//
//  Reminder+Create.swift
//  SmartPill
//
//  Created by Dennis da Silva Nunes on 20/05/15.
//  Copyright (c) 2015 Dennis da Silva Nunes. All rights reserved.
//

import UIKit
import CoreData

extension Reminder{
    
    class func createReminder(medicine:Medicine?,remSchedule:Reminder_Schedule,remSound:Reminder_Sound?)->Reminder?{
        
        let appDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = appDelegate.managedObjectContext
        
        if((medicine) != nil){
            let medicineName = medicine?.name
            
            let request = NSFetchRequest(entityName: "Medicine")
            request.predicate = NSPredicate(format: "name = %@", medicineName!)
            
            var matches = context!.executeFetchRequest(request, error: nil) as? [Medicine]
            if(matches == nil || matches!.count > 1){
                return nil
            }else if(matches!.count==1){
                for m in matches! {
                    if(m.name == medicineName){
                    var rem = NSEntityDescription.insertNewObjectForEntityForName("Reminder", inManagedObjectContext: context!) as! Reminder
                    rem.medicine = m
                    rem.reminder_schedule = remSchedule
                    if remSound != nil {rem.reminder_sound = remSound!}
                    context?.save(nil)
                    return rem
                    }
                }
            }else{
                return nil
            }
        }
        return nil
    }
}